### PR TITLE
major refactor of the module for terraform 0.13

### PR DESCRIPTION
**Pull request by dennybaa**
* breaks compatibility with the previous versions

### CODE REVIEW

**Comment by comment-bot-dev** [comment]
Thanks for the PR! 🚀
Unfortunately it looks like some of our CI checks failed. See the [Contributing Guide](../blob/master/CONTRIBUTING.md) for details.
- ⚠️check_documentation
 The documentation needs to be regenerated. Please run `make generate_docs`. 
 ```bash 
Checking for documentation generation
diff -r '--exclude=.terraform' '--exclude=.kitchen' '--exclude=.git' '--exclude=autogen' '--exclude=*.tfvars' /workspace/README.md /tmp/tmp.Qo8IPuHKy6/generate_docs/workspace/README.md
60c60
< | key\_opts\_default | Specifies the default key options (override with cation). | `any` | <pre>{<br>  "algorithm": "GOOGLE_SYMMETRIC_ENCRYPTION",<br>  "labels": {},<br>  "prevent_destroy": true,<br>  "protection_level": "SOFTWARE",<br>  "rotation_period": "100000s"<br>}</pre> | no |
---
> | key\_opts\_default | Specifies the default key options (override with caution). | `any` | <pre>{<br>  "algorithm": "GOOGLE_SYMMETRIC_ENCRYPTION",<br>  "labels": {},<br>  "prevent_destroy": true,<br>  "protection_level": "SOFTWARE",<br>  "rotation_period": "100000s"<br>}</pre> | no |
diff -r '--exclude=.terraform' '--exclude=.kitchen' '--exclude=.git' '--exclude=autogen' '--exclude=*.tfvars' /workspace/examples/simple_example/README.md /tmp/tmp.Qo8IPuHKy6/generate_docs/workspace/examples/simple_example/README.md
9,13c9,13
< |------|-------------|:----:|:-----:|:-----:|
< | keyring | Keyring name. | string | n/a | yes |
< | keys | Key names. | list(string) | `<list>` | no |
< | location | Location for the keyring. | string | `"global"` | no |
< | project\_id | The ID of the project in which to provision resources. | string | n/a | yes |
---
> |------|-------------|------|---------|:--------:|
> | keyring | Keyring name. | `string` | n/a | yes |
> | keys | Key names. | `list(string)` | `[]` | no |
> | location | Location for the keyring. | `string` | `"global"` | no |
> | project\_id | The ID of the project in which to provision resources. | `string` | n/a | yes |
Error: Documentation generation has not been run, please run the
'make docker_generate_docs' command and commit the above changes.
 ``` 
- ⚠️check_headers
 All files need a license header. Please make sure all your files include the appropriate header. A helper tool available [here](https://github.com/google/addlicense). 
 ```bash 
Checking file headers
3 files have incorrect boilerplate headers:
locals.tf
main.tf
versions.tf
 ``` 
- ⚠️check_terraform
 Failed Terraform check. More details below. 
 ```bash 
Running terraform fmt
Running terraform validate
terraform_validate . 
Success!
 The configuration is valid.
terraform_validate ./examples/simple_example 
Error: 
Unsupported argument
  on main.tf line 26, in module "kms":
  26:   
key_opts
 = [
An argument named "key_opts" is not expected here.
terraform_validate ./test/fixtures/simple_example 
Warning: 
Interpolation-only expressions are deprecated
  on outputs.tf line 34, in output "project_id":
  34:   value       = 
"${var.project_id}"
Terraform 0.11 and earlier required all non-constant expressions to be
provided via interpolation syntax, but this pattern is now deprecated. To
silence this warning, remove the "${ sequence from the start and the }"
sequence from the end of this expression, leaving just the inner expression.
Template interpolation syntax is still used to construct strings from
expressions when the template includes multiple interpolation sequences or a
mixture of literal strings and interpolations. This deprecation applies only
to templates that consist entirely of a single interpolation sequence.
Error: 
Unsupported argument
  on ../../../examples/simple_example/main.tf line 26, in module "kms":
  26:   
key_opts
 = [
An argument named "key_opts" is not expected here.
terraform_validate ./test/setup 
Error: 
Unsupported Terraform Core version
  on .terraform/modules/project_ci_kms/versions.tf line 18, in terraform:
  18:   
required_version = "~> 0.12.6"
Module module.project_ci_kms (from
terraform-google-modules/project-factory/google) does not support Terraform
version 0.13.5. To proceed, either choose another supported Terraform version
or update this version constraint. Version constraints are normally set for
good reason, so updating the constraint may lead to other errors or unexpected
behavior.
Error: 
Unsupported Terraform Core version
  on .terraform/modules/project_ci_kms/modules/budget/versions.tf line 18, in terraform:
  18:   
required_version = "~> 0.12.6"
Module module.project_ci_kms.module.budget (from ./modules/budget) does not
support Terraform version 0.13.5. To proceed, either choose another supported
Terraform version or update this version constraint. Version constraints are
normally set for good reason, so updating the constraint may lead to other
errors or unexpected behavior.
Error: 
Unsupported Terraform Core version
  on .terraform/modules/project_ci_kms/modules/core_project_factory/versions.tf line 18, in terraform:
  18:   
required_version = "~> 0.12.6"
Module module.project_ci_kms.module.project-factory (from
./modules/core_project_factory) does not support Terraform version 0.13.5. To
proceed, either choose another supported Terraform version or update this
version constraint. Version constraints are normally set for good reason, so
updating the constraint may lead to other errors or unexpected behavior.
Error: 
Unsupported Terraform Core version
  on .terraform/modules/project_ci_kms/modules/gsuite_group/versions.tf line 18, in terraform:
  18:   
required_version = "~> 0.12.6"
Module module.project_ci_kms.module.gsuite_group (from ./modules/gsuite_group)
does not support Terraform version 0.13.5. To proceed, either choose another
supported Terraform version or update this version constraint. Version
constraints are normally set for good reason, so updating the constraint may
lead to other errors or unexpected behavior.
Error: 
Unsupported Terraform Core version
  on .terraform/modules/project_ci_kms/modules/project_services/versions.tf line 18, in terraform:
  18:   
required_version = "~> 0.12.6"
Module module.project_ci_kms.module.project-factory.module.project_services
(from ../project_services) does not support Terraform version 0.13.5. To
proceed, either choose another supported Terraform version or update this
version constraint. Version constraints are normally set for good reason, so
updating the constraint may lead to other errors or unexpected behavior.
 ```

File: locals.tf [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+    ]
+  ])
+
+  owners_list = flatten([
+    for rule in var.acl : [
+      for identity in lookup(rule, "owners", []) : {
+        key      = rule.key
+        key_id   = local.crypto_keys[rule.key].id
+        identity = identity
+        role     = "roles/cloudkms.cryptoKeyEncrypter"
```
Comment (bharathkkb) [inline locals.tf:59]:
```suggestion
        role     = "roles/owner"
```

File: examples/simple_example/main.tf [hunk 2]
```diff
@@ -14,17 +14,20 @@
  * limitations under the License.
  */
 
-provider "google" {
-  version = "~> 3.0.0"
-}
-
 module "kms" {
-  source     = "../.."
+  source  = "terraform-google-modules/kms/google"
```
Comment (bharathkkb) [inline examples/simple_example/main.tf:18]: We use the examples for testing, so we should continue to keep these as `../..`

File: main.tf [hunk 3]
```diff
@@ -1,83 +1,67 @@
-/**
```
Comment (bharathkkb) [inline main.tf:1]: headers should be preserved

File: main.tf [hunk 4]
```diff
@@ -1,83 +1,67 @@
...
- */
 
-locals {
-  keys_by_name = zipmap(var.keys, var.prevent_destroy ? slice(google_kms_crypto_key.key[*].self_link, 0, length(var.keys)) : slice(google_kms_crypto_key.key_ephemeral[*].self_link, 0, length(var.keys)))
+provider "google" {
+  version = "< 4.0"
 }
 
-resource "google_kms_key_ring" "key_ring" {
+/* provider "google-beta" {
```
Comment (bharathkkb) [inline main.tf:6]: nit: commented code

File: README.md [hunk 5]
```diff
@@ -27,10 +25,23 @@ module "kms" {
   location           = "europe"
   keyring            = "sample-keyring"
   keys               = ["foo", "spam"]
-  set_owners_for     = ["foo", "spam"]
-  owners = [
-    "group:one@example.com,group:two@example.com",
-    "group:one@example.com",
+
+  key_opts = [
```
Comment (bharathkkb) [inline README.md:29]: IIUC, this for overriding `key_opts_default` for individual keys, maybe we can simplify this to a map like keyName => {options}. WDYT?
```tf
key_opts = [ 
  foo = {rotation_period = "100000s"},
  bar = {algorithm = "EXTERNAL_SYMMETRIC_ENCRYPTION"}
]
```

**Review by bharathkkb — CHANGES_REQUESTED** [review]
Thanks for the PR @dennybaa

**Comment by morgante** [comment]
@dennybaa Can you update based on feedback? We'd like to include this in the next release.

**Comment by dennybaa** [comment]
@morgante Hey. Yup I will look at it the other day.

**Comment by morgante** [comment]
@dennybaa Gentle ping, are you still interested in working on this?

**Comment by dennybaa** [comment]
@morgante Howdy. It took a while, nonetheless I have rebased. Plz have a look.

File: README.md (hunk 5 above)
Comment (dennybaa) [reply README.md:31]: Doesn't simplify much, and I prefer more the behavior seen in google kubernetes module...

File: locals.tf (hunk 1 above)
Comment (dennybaa) [reply locals.tf]: well this's wrong. making the identity owner of the project it's not the intent... Owner here is the "owner" of a key, i.e.  encrypter/decrypter.

File: examples/simple_example/main.tf (hunk 2 above)
Comment (dennybaa) [reply examples/simple_example/main.tf]: fixed

File: main.tf (hunk 4 above)
Comment (dennybaa) [reply main.tf]: removed

File: locals.tf (hunk 1 above)
Comment (dennybaa) [reply locals.tf]: Well I guess. You're right looked more carefully how IAM works in this regard.

**Comment by morgante** [comment]
Test failure:
```
>>>>>> Input 'keys' with value '{:foo=>"projects/ci-kms-module-007c/locations/global/keyRings/sample-keyring/cryptoKeys/foo"}' does not validate to type 'Array'.
````

**Comment by dennybaa** [comment]
image

It would be nice @morgante , if you could take from here.

**Comment by morgante** [comment]
@dennybaa Have you tried getting the [linters](https://github.com/terraform-google-modules/terraform-google-kms/blob/master/CONTRIBUTING.md#linting-and-formatting) to pass? That should at least be easy to do locally.

**Comment by dennybaa** [comment]
> @dennybaa Have you tried getting the [linters](https://github.com/terraform-google-modules/terraform-google-kms/blob/master/CONTRIBUTING.md#linting-and-formatting) to pass? That should at least be easy to do locally.

@morgante noop. I would prefer to have visibility of the issues from the UI. Unfortunately I have no possibility to invest into a full-fledged investigation. Yeah especially when bringing the whole kitchen up...

**Comment by morgante** [comment]
@dennybaa Linter failures are commented here: https://github.com/terraform-google-modules/terraform-google-kms/pull/37#issuecomment-787033441

**Comment by dennybaa** [comment]
@morgante haha:) my bad. Thanks for pointing out.

**Comment by dennybaa** [comment]
@morgante gentle ping. Could you please share the details of the failure, I think I don't see them...

**Comment by morgante** [comment]
This is the error:
```
$$$$$$ Verifying the 'simple_example local' system...
>>>>>> Input 'keys' with value '{:foo=>"projects/ci-kms-module-7127/locations/global/keyRings/sample-keyring/cryptoKeys/foo"}' does not validate to type 'Array'.
```

For the examples, we should keep the outputs the same: https://github.com/terraform-google-modules/terraform-google-kms/pull/37/files#diff-866254505b3bdeb20c427890799fad5fdbebbfbbeaf9cf8c1f4e0f51fab9be82R29

**Comment by dennybaa** [comment]
> This is the error:
> 
> ```
> $$$$$$ Verifying the 'simple_example local' system...
> >>>>>> Input 'keys' with value '{:foo=>"projects/ci-kms-module-7127/locations/global/keyRings/sample-keyring/cryptoKeys/foo"}' does not validate to type 'Array'.
> ```
> 
> For the examples, we should keep the outputs the same: https://github.com/terraform-google-modules/terraform-google-kms/pull/37/files#diff-866254505b3bdeb20c427890799fad5fdbebbfbbeaf9cf8c1f4e0f51fab9be82R29

@morgante Thanks. fixed, but smth else.

**Comment by morgante** [comment]
@dennybaa This is the error now:

```
Profile Summary: 1 successful control, 0 control failures, 0 controls skipped
Test Summary: 6 successful, 0 failures, 0 skipped
$$$$$$ Finished verifying the 'simple_example local' system.
$$$$$$ Verifying the 'simple_example gcp' system...
>>>>>> ------Exception-------
>>>>>> Class: Kitchen::ActionFailed
>>>>>> Message: 1 actions failed.
>>>>>>     Failed to complete #verify action: [cannot load such file -- google/apis/dns_v2beta1] on simple-example-default
```

We might need to update the inspec-gcp dependency to the latest: https://github.com/terraform-google-modules/terraform-google-kms/blob/master/test/integration/simple_example/inspec.yml#L4

**Comment by dennybaa** [comment]
@morgante hey. a friendly reminder.

File: README.md (hunk 5 above)
Comment (morgante) [reply README.md:31]: This field seems to be redundant with `keys`. I suggest we either:

1. Make `keys` a list of objects (including name and optional parameters).
2. Make this into a map of `key_name -> object()`.

File: examples/simple_example/main.tf [hunk 6]
```diff
@@ -19,12 +19,25 @@ provider "google" {
...
 
+  key_opts = [
+    {
+      key             = "foo"
+      rotation_period = "100000s"
+      prevent_destroy = false
+    },
+  ]
+
+  # acl = [
```
Comment (morgante) [inline examples/simple_example/main.tf:37]: We should not have commented out code.

File: variables.tf [hunk 7]
```diff
@@ -41,61 +47,26 @@ variable "prevent_destroy" {
...
+  description = "Access control list for the managed keys."
+  type        = list(any)
   default     = []
 }
 
-variable "set_decrypters_for" {
-  description = "Name of keys for which decrypters will be set."
-  type        = list(string)
-  default     = []
+variable "key_opts_default" {
```
Comment (morgante) [inline variables.tf:56]: Collapsing many options into a single variable is an anti-pattern, we should continue to leave these as top-level variables.

**Review by morgante — CHANGES_REQUESTED** [review]

**Pull request closed without merging**